Image Loading Optimisation Module

kt-12 opened the pull request:
## PR Checklist
<!-- 
Please check if your PR fulfills the following requirements:
**Filling out the template is required.** Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
 -->
- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md


## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x" with no spaces eg: [x]. -->
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

kt-12 committed:
LoadingOptimization Module

kt-12 committed:
change filter

kt-12 committed:
cs fixes

kt-12 committed:
remove filter and replace with variables

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,264 @@
+	public function __construct() {
+		parent::__construct();
+
+		add_action( 'wp_get_attachment_image_attributes', [ $this, 'remove_get_attachment_loading_attributes' ], 10, 3 );
```
The two hooks: `wp_get_attachment_image_attributes` and `wp_content_img_tag` are both filters rather than actions. should be using `add_filter()` instead.

**Also, a side note:**

I'd generally suggest trying to avoid setting hooks in an object constructor, as this creates side effects whenever the object is instantiated. A better pattern is to include in some sort of a setup method that is called after the object is instantiated, like this (pseudocode example):

```php
$module = new Module();
$module->setup(); // Responsible for registering hooks, registering options, etc.
```
The advantage to this type of pattern is that it separates the object state from its interaction with the application, making it easier to test in isolation. For more, see: https://tommcfarlin.com/wordpress-plugin-constructors-hooks/.

However, calling hooks in module constructors seems to be a common pattern in this codebase, so no need to make that change here.

kt-12 replied:
I was aware of the side effect and I usually don't prefer calling hooks in constructor, but the way dynamic loading of the modules are written here, it was the only option to call hooks. Also, since this was something they have followed for all the modules I didn't spend much time on this part.

I have fix `action` to `filter`

joemcgill replied:
Assumed you were probably aware, but it wasn't something I remember us discussing before, so figured I'd mention it. Thanks!

kt-12 replied:
It's good that you raised it. There are many improvements that can be made in elementor beyond the scope of this work. Something that we can note and pickup later.

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,264 @@
+		add_filter( 'the_content', [ $this, 'filter_images' ], 10, 1 );
+	}
+
+	public function remove_get_attachment_loading_attributes( $attr, $attachment, $size ) {
+		unset( $attr['fetchpriority'] );
+		unset( $attr['loading'] );
+		return $attr;
+	}
+
+	public function remove_content_img_tag_loading_attributes( $filtered_image, $context, $attachment_id ) {
+		$tag = new \WP_HTML_Tag_Processor( $filtered_image );
+		if ( $tag->next_tag() ) {
+			$tag->remove_attribute( 'fetchpriority' );
+			$tag->remove_attribute( 'loading' );
+		}
+		return $tag->get_updated_html();
+	}
```
Both of these filters remove core's attributes after they are applied, which gets the desired result, but means that unnecessary code is running to calculate these values before discarding them. Instead, I wonder if we could do something like the following to keep core from applying these attributes in the first place?

```php
add_filter( 'wp_lazy_loading_enabled', '__return_false' );
add_action( 'init', function () {
	wp_high_priority_element_flag( false );
} );
```

kt-12 replied:
Yes this make sense. I have added it  to the constructor.

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,264 @@
+		foreach ( $matches as $match ) {
+			$tag = $match[0];
+			if ( preg_match( '/wp-image-([0-9]+)/i', $tag, $class_id ) ) {
+					$attachment_id = absint( $class_id[1] );
```
Tabs are off here.

kt-12 replied:
Fixed, not sure why this got skipped by phpcs.

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,264 @@
+					/*
+						* If exactly the same image tag is used more than once, overwrite it.
+						* All identical tags will be replaced later with 'str_replace()'.
+						*/
```
Tab spacing here too.

kt-12 replied:
Fixed

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,264 @@
+		return $tag->get_updated_html();
+	}
+
+	public function filter_images( $content ) {
```
Rather than redoing the `preg_match` work from `wp_filter_content_tags`, could we hook into `wp_content_img_tag` and only run the logic you've applied in the final foreach loop to add our own optimizations?

kt-12 replied:
Thanks for pointing this out, it saved a lot of computing. Testing this out thoroughly before I commit the change.

joemcgill reviewed (commented):
The architecture of this is looking really good! I've some initial ideas/comments about specifics of the implementation inline.

I think it would be good to take some time early on to create some test cases for this module, based mostly on the tests for the related functionality in core, to make sure this is working as expected. However, one specific set of tests we need to create here are to mock the metadata that is generated by elementor for things like:

* A post with an image widget
* A post with an image widget before a gallery widget
* A post with a gallery widget before an image
* (etc.)

This will help us ensure that we're getting the right output as we iterate on the implementation.

joemcgill commented:
Spent a bit of time thinking about how we could handle header images. By default, Elementor loads the header and footer of a page using the `header-footer.php` template in `modules/page-templates/templates/header-footer.php`. This file calls `get_header()`to display the active theme's `header.php` file. We could consider firing an action on `get_header` that starts an output buffer and then another on the `elementor/page_templates/header-footer/before_content` action that runs post processing on the markup before echoing it out. What do you think?

kt-12 committed:
add_action to filter & add early exit of logic

kt-12 committed:
remove unwanted complexity and also disable core lazy loading and fetchpriority

kt-12 committed:
update codebase

kt-12 committed:
introduce new action to know footer has ended.

kt-12 committed:
filter header and footer

kt-12 commented:
> elementor/page_templates/header-footer/before_content

I have tried this approach and it seems to work fine. However I am not sure what if someone has implemented buffer within any of the action.

kt-12 committed:
reusing function

kt-12 committed:
remove extra line

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,299 @@
+			/*
+			* Warm the object cache with post and meta information for all found
+			* images to avoid making individual database calls.
+			*/
```
Multi-line docs should have the `*` aligned with a leading space.

```suggestion
			 * Warm the object cache with post and meta information for all found
			 * images to avoid making individual database calls.
			 */
```

kt-12 replied:
Fixed.

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,299 @@
+				/*
+				* Unset image lookup to not run the same logic again unnecessarily if the same image tag is used more than
+				* once in the same blob of content.
+				*/
```
Same as above

```suggestion
				/*
				 * Unset image lookup to not run the same logic again unnecessarily if the same image tag is used more than
				 * once in the same blob of content.
				 */
```

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,299 @@
+		// wp_high_priority_element_flag was only introduced in 6.3.0
+		if( function_exists('\wp_high_priority_element_flag') ) {
+			\wp_high_priority_element_flag( false );
+		}
```
✨ Good use of function_exists here.

Coding style changes:
* Spacing
* No need to qualify WP functions with a leading slash

```suggestion
		if ( function_exists( 'wp_high_priority_element_flag' ) ) {
			wp_high_priority_element_flag( false );
		}
```

kt-12 replied:
Fixed.

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,299 @@
+		}
+	}
+
+	public function remove_get_attachment_loading_attributes( $attr, $attachment, $size ) {
```
If we're only using the first parameter here, the others can be removed.

```suggestion
	public function remove_get_attachment_loading_attributes( $attr ) {
```

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,299 @@
+		add_action( 'elementor/page_templates/header-footer/before_content', [ $this, 'flush_buffer' ] );
+
+		// Run optimization logic on content.
+		add_filter( 'wp_get_attachment_image_attributes', [ $this, 'remove_get_attachment_loading_attributes' ], 10, 3 );
```
I'm curious why this is needed, since we're disabling these attributes in lines 25 and 26 above? Also, nitpick: since the callback is only using the first parameter, this could be simplified to:
```suggestion
		add_filter( 'wp_get_attachment_image_attributes', [ $this, 'remove_get_attachment_loading_attributes' ] );
```

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,299 @@
+		$filtered_image = $this->add_loading_optimization_attrs( $filtered_image );
+		self::$image_visited[ $attachment_id ] = $filtered_image;
+		return $filtered_image;
+	}
```
This is a smart idea as an optimization, but currently could result in two different sizes for the same image being given the same optimization unnecessarily. For example if the thumbnail and full size file for the same attachment_id were shown on the same page, the loading attributes for the first would be applied to the second, which is not what we want.

This was also confusing me while doing functional tests because I had the same image showing multiple times in the same post and was seeing unexpected results.

For now, I think we can remove this and hook `add_loading_optimization_attrs` directly to the `wp_content_img_tag` filter in line 34, the `$image_visited` property can also be removed since it will no longer be in use.

kt-12 replied:
Thanks for pointing this out. I was trying to adopt optimisation strategy from `wp_filter_content_tags` however I add check for `attachement_id` instead of the tag. I have fixed it now. Now two image will get same strategy if we have exactly same tags (something that currently happens in core). This makes sense as images are not loaded again by the browser. 

The only thing to test is if there any side-effect to this optimisation. For example, if we have 10 large images (only) with similar tag then all the 10 images will be marked `fetchpriority=high`, w.r.t loading image, image will be loaded only once, but the presence of `fetchpriority=high` on other copies may/or may not add slight delay (this is something that needs to be lab tested).

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,299 @@
+				_doing_it_wrong(
+					__FUNCTION__,
+					__( 'An image should not be lazy-loaded and marked as high priority at the same time.', 'elementor' )
+				);
```
_doing_it_wrong() requires a third param for `$version` (related to the WP version where the message was added). For plugins, you can just pass an empty string. 

```suggestion
				_doing_it_wrong(
					__FUNCTION__,
					__( 'An image should not be lazy-loaded and marked as high priority at the same time.', 'elementor' ),
					''
				);
```

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,299 @@
+		}
+
+		if ( $maybe_in_viewport ) {
+			$loading_attrs = $this->maybe_add_fetchpriority_high_attr( $loading_attrs, $tag_name, $attr );
```
`$tag_name` is undefined here. Since we're primarily dealing with images here, perhaps we should remove this param entirely until we add additional logic for iframes?

kt-12 replied:
Fixed.

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,299 @@
+		return $omit_threshold;
+	}
+
+	private function maybe_add_fetchpriority_high_attr( $loading_attrs, $tag_name, $attr ) {
```
Consider removing the unused `$tag_name` param, per my prior comment

```suggestion
	private function maybe_add_fetchpriority_high_attr( $loading_attrs, $attr ) {
```

kt-12 replied:
Fixed.

joemcgill reviewed (dismissed):
Thanks @kt-12. The updates are looking pretty good. I've left another round of feedback on some things that need to be addressed. Additionally, docblocks need to be added throughout.

I'm also thinking we'll need to ad a method [like this one](https://github.com/elementor/elementor/blob/main/modules/lazyload/module.php#L21-L30), so this is loaded as an experimental module—along with adding a class or data-attribute to the markup so they can detect whether a site has the experiment enabled. We can get specific guidance from them about to set that up.

kt-12 committed:
remove defaults

Co-authored-by: Joe McGill <801097+joemcgill@users.noreply.github.com>

kt-12 committed:
add empty string

Co-authored-by: Joe McGill <801097+joemcgill@users.noreply.github.com>

kt-12 committed:
fix cbf issue and update optimization code.

kt-12 committed:
change function name to more relevant context.

kt-12 committed:
optimize code further, don't run filter if the image was already loaded.

kt-12 committed:
remove unwanted optimization

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,264 @@
+
+			// Filter an image match.
+			$filtered_image = $tag;
+			$attachment_id  = $images[ $tag ];
```
`$images` is undefined here. In `wp_filter_content_tags` all the matches get collected to an array of attachment IDs in the first loop so they can all be run through `_prime_post_caches()`, which is what this line is referencing.  Since we're not worried about fetching any attachment data in this loop, we can remove this line entirely.

kt-12 replied:
@joemcgill I have modified the code to remove `_prime_post_caches`  and made some more changes to the function to optimise it.

kt-12 committed:
Docblock + optimization + fixing header logic

kt-12 committed:
remove the addition hook

kt-12 committed:
remove additional comment sign

kt-12 committed:
change priority to 0

kt-12 committed:
doc block for variables

kt-12 committed:
rename function

kt-12 committed:
update doc block

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,333 @@
+	 * @return string We simply return the content from parameter.
+	 */
+	public function flush_header_buffer( $content ) {
+		$this->flush_buffer();
```
I don't think another method is needed here, you can probably call `ob_end_flush()` directly here and remove the `flush_buffer` method.

kt-12 replied:
@joemcgill I have added some new logic to check status before flushing, so retained `flush_buffer()`

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,333 @@
+		// Run optimization logic on header.
+		add_action( 'get_header', [ $this, 'set_buffer' ] );
+		// Ensure buffer is flushed (if any) before the content logic.
+		add_filter( 'the_content', [ $this, 'flush_header_buffer' ], 0 );
```
Clever idea. So no matter which template is in use, output buffering will collect anything prior to the first time `the_content` is called, and this will cause all of that output to be flushed before moving forward with our content processing. On a page like an archive where `the_content` is called multiple times in the same page, we may want to make sure that this doesn't cause any side effects. We could potentially add a check for `ob_get_status` prior to flushing and only do so if there is an active output buffer with `handle_buffer_content` set as the name.

westonruter replied:
Yeah, it's also important to make sure that the flushing doesn't accidentally flush another output buffer that another plugin may have opened (e.g. a caching plugin). I suppose that probably won't be the case since `get_header` is much later than caching plugins start their output buffering. But another plugin could start an output buffer at the `wp_head` action, in which case the call to `ob_end_flush()` won't flush the right buffer: it would flush the outermost buffer(s) not the one that was started at `get_header`. To do that, there would probably need to be a loop that checks `ob_get_status()` until it gets to `handle_buffer_content` as @joemcgill suggests.

Nevertheless, I wonder if this will be free of all side effects, as again it would cause buffers to get flushed perhaps prematurely for some plugins. 

What if instead of buffering just the header and just the footer, it instead buffers the whole page? Then it could inject some HTML comment markers for where `the_content` first starts and last ends, and then the output buffer processing can just process everything up to the first comment (the header) and everything after the last (the footer)?

kt-12 replied:
@joemcgill Thanks for pointing this out. I have added the check [here](https://github.com/kt-12/elementor/pull/1/files#diff-b35207b5df0f5802570d77f8eb0f6b5d697b911ba1049355b6ff34a42b55de24R79). 
@westonruter you have a valid point, to test this I did previously set `ob_start` in `init` and let PHP end script take the default course and call the handler. 
It did work fine but I didn't go that path for couple of reason (just my theory need to validate it through lab test).

1) print  would only happen towards the end, so I wanted to see if it effects TTFB of a site as compared to the same site printing out html in parts. I did read from some stackoverflow(not so authentic claim) that they had issues of page loading slow due to full page buffering they implemented.
2) Need to check effects of full page buffering on memory -> considering the fact that there can be sites that run on low-end servers like `aws lightsail $3.50` or `$5` servers. 
I feel full page buffer must go in combination with a good caching logic.
3) I wasn't sure what happens to a large page with a lower buffer size - will the the flush happen in between and handle gets buffer in chunks instead of a single buffer. 
4) Also we could avoid a loop when we use `wp_content_img_tag` as compared to buffer - content would have most number of images for majority of sites.

joemcgill replied:
Thanks for the update. Ideally, think Weston's right, that it would be best to process the whole page in a single buffer, which was our original objective. Though I agree with you that finding an approach that reliably captures all of the page output, including any custom theme logic that happens prior to Elementor's output filtering, and does not interfere with any output buffering done by third parties, is a tricky problem to solve. Let's keep thinking about how to make the approach more robust, but I don't see that as an immediate blocker for now.

joemcgill reviewed (commented):
Some minor notes, but this looks good to me as soon as we've got some unit tests added. Thanks, @kt-12

westonruter commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,333 @@
+	 * @param string $buffer Buffered content.
+	 * @return string Content with optimized images.
+	 */
+	public function handel_buffer_content( $buffer ) {
```
Nit:

```suggestion
	public function handle_buffer_content( $buffer ) {
```

kt-12 replied:
Fixed.

westonruter commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,333 @@
+		parent::__construct();
+
+		$this->min_priority_img_pixels = 50000;
+		$this->omit_threshold = 3;
```
Any reason to not just set these above when defining the private member variables?

westonruter commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,333 @@
+	 * @return string Content with optimized images.
+	 */
+	private function filter_images( $content ) {
+		if ( ! preg_match_all( '/<img\s[^>]+>/', $content, $matches, PREG_SET_ORDER ) ) {
+			return $content;
+		}
+
+		// Iterate through the matches in order of occurrence as it is relevant for whether or not to lazy-load.
+		foreach ( $matches as $match ) {
+			$tag = $match[0];
+
+			// Optimize an image.
+			$optimized_image = $this->loading_optimization_image( $tag );
+
+			if ( $optimized_image !== $tag ) {
+				$content = str_replace( $tag, $optimized_image, $content );
+			}
+		}
```
Instead of using `preg_match_all()` followed by `str_replace()`, I recommend instead a single call to `preg_replace_callback()`. This is much faster as it avoids re-searching throughout the content many times. This was an issue in core as well which I filed and fixed in https://core.trac.wordpress.org/ticket/58069

kt-12 replied:
Thank you @westonruter. This really made sense, it also helped cut down some code.

kt-12 committed:
Enable fetchpriority for is_admin too

kt-12 committed:
test fetch priority

kt-12 committed:
buffer callback check before flushing

kt-12 committed:
use empty instead of isset

kt-12 committed:
remove flush function

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,326 @@
+		}
+
+		if ( null === $maybe_in_viewport ) {
+			$content_media_count = $this->increase_content_media_count( 0 );
```
Let's add a `@todo` inline comment to note that the in_admin check was removed to accommodate unit tests.

kt-12 replied:
added

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,326 @@
+		$buffer_status = ob_get_status();
+		if ( ! empty( $buffer_status ) &&
+			1 === $buffer_status['type'] &&
+			get_class( $this ) . '::handle_buffer_content' == $buffer_status['name'] ) {
```
👍🏻 I think this will be safer than a generic flush, though it could lead to edge cases where this gets skipped and the output buffer is not flushed at all. We should do some testing with different scenarios where a second buffer is started before and/or after ours, just in case.

Nit: use strict comparison.

```suggestion
			get_class( $this ) . '::handle_buffer_content' === $buffer_status['name'] ) {
```

joemcgill commented on tests/phpunit/elementor/modules/image-loading-optimization/test-module.php:
```diff
@@ -0,0 +1,39 @@
+		// Arrange.
+		/** @var PageTemplatesModule $page_templates_module */
+		$page_templates_module = Plugin::$instance->modules_manager->get_modules( 'page-templates' );
+		$image_loading_optimization_module = Plugin::$instance->modules_manager->get_modules( 'image-loading-optimization' );
```
This isn't being used. Is it needed to ensure the module is loaded?

joemcgill commented on tests/phpunit/elementor/modules/image-loading-optimization/test-module.php:
```diff
@@ -0,0 +1,39 @@
+		// Simulate a singular query.
+		query_posts( [ 'p' => $document->get_main_id() ] );
+		ob_start();
+		$page_templates_module->print_content();
```
This method basically just runs `the_content()` as long as there are posts in the query to render ([code ref](https://github.com/kt-12/elementor/blob/main/modules/page-templates/module.php#L179-L192)), so this will test the post content, but not any header or footer logic. This is fine for a very basic test case, but we'll want to add additional test cases that include header and footer content. It also makes me curious if the `_wp_page_template` metadata you're setting on this post is having any affect in this test case?

joemcgill reviewed (commented):
Thanks, @kt-12. A few nitpicks, otherwise this is getting close.

kt-12 committed:
Update modules/image-loading-optimization/module.php

Co-authored-by: Joe McGill <801097+joemcgill@users.noreply.github.com>

kt-12 committed:
update logic with preg_replace_callback

kt-12 committed:
Merge branch 'feature/img-lo-wp-logic-to-elementor' of https://github.com/kt-12/elementor into feature/img-lo-wp-logic-to-elementor

westonruter commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,321 @@
+	 * @return string Content with optimized images.
+	 */
+	private function filter_images( $content ) {
+		return preg_replace_callback( '/<img\s[^>]+>/', [ $this, 'apply_image_optimization' ], $content );
+	}
+
+	/**
+	 * Callback to apply optimization logic on the image.
+	 * This function is called for each image tag found in the content.
+	 *
+	 * @param array $matches Array of matches.
+	 * @return string Optimized image.
+	 */
+	private function apply_image_optimization( $matches ) {
+		return $this->loading_optimization_image( $matches[0] );
```
This may be a bit less verbose (since the `apply_image_optimization` method is only used once) and probably easier to understand:

```suggestion
		return preg_replace_callback(
			'/<img\s[^>]+>/',
			function ( $matches ) {
				return $this->loading_optimization_image( $matches[0] )
			},
			$content
		);
```

kt-12 committed:
test with buffer issue

kt-12 committed:
remove print

kt-12 committed:
path issue fix

kt-12 committed:
handling buffer issue for template.

kt-12 committed:
buffer fix

kt-12 committed:
Merge pull request #2 from kt-12/feature/temp-test

Fix test case with buffer issue.

kt-12 committed:
remove additional function

Co-authored-by: Weston Ruter <westonruter@google.com>

kt-12 committed:
syntax ending issue fix

joemcgill reviewed (commented):
The latest updates are looking really good, and it's nice to see the tests passing now for me locally. Thanks @kt-12! 

Unless I'm mistaken, I think all that's left here is to add a `@todo` about fixing the `! is_admin()` logic and adding support for this module to be experimental and we can wrap this one up.

kt-12 committed:
Add experimental setting

kt-12 committed:
remove extra line

kt-12 committed:
add todo

kt-12 commented:
> Remember, contributions to this repository should follow our [GitHub Community Guidelines](https://docs.github.com/articles/github-community-guidelines).

Added both the changes for review.

joemcgill commented on modules/image-loading-optimization/module.php:
```diff
@@ -0,0 +1,340 @@
+
+		// Run optimization logic on footer. Flushing of footer buffer will be handled by PHP script end default logic.
+		add_action( 'get_footer', [ $this, 'set_buffer' ] );
+		add_action( 'shutdown', [ $this, 'set_buffer' ] );
```
Was this leftover from testing or is there another reason for this action?

kt-12 replied:
Sorry that was a left over. Have removed it.

joemcgill commented on tests/phpunit/elementor/modules/image-loading-optimization/test-module.php:
```diff
@@ -0,0 +1,68 @@
+<?php
+namespace Elementor\Tests\Phpunit\Elementor\Modules\Image_Loading_Optimization;
+
+use Elementor\Core\Base\Document;
+use Elementor\Modules\PageTemplates\Module as PageTemplatesModule;
+use Elementor\Plugin;
+use ElementorEditorTesting\Elementor_Test_Base;
+use Elementor\TemplateLibrary\Source_Local;
```
A few of these imported classes don't seem to be used. Can we remove them? Alternately, it looks like the places where we are using the modules manager to get the `page-templates` module, we could instead use the `PageTemplatesModule` class that is being imported here and avoid some undefined method and undefined class constant warnings.

```suggestion
use Elementor\Plugin;
use ElementorEditorTesting\Elementor_Test_Base;
```

kt-12 replied:
Fixed.

joemcgill commented on tests/phpunit/elementor/modules/image-loading-optimization/test-module.php:
```diff
@@ -0,0 +1,68 @@
+	}
+
+	public function get_page_template() {
+		$page_templates_module = Plugin::$instance->modules_manager->get_modules( 'page-templates' );
+		return [
+			[ $page_templates_module::TEMPLATE_CANVAS ],
+			[ $page_templates_module::TEMPLATE_HEADER_FOOTER ],
+			[ $page_templates_module::TEMPLATE_THEME ]
+		];
```
An example of using the imported class here:

```suggestion
		return [
			[ PageTemplatesModule::TEMPLATE_CANVAS ],
			[ PageTemplatesModule::TEMPLATE_HEADER_FOOTER ],
			[ PageTemplatesModule::TEMPLATE_THEME ]
		];
```

joemcgill reviewed (commented):
Nice updates! A few nitpicks, but I think this is generally ready to get external feedback on. I'll open an issue on the Elementor repo and then we can close this PR in favor of a new one directly to their repo

kt-12 committed:
remove unused namespaces

joemcgill commented:
@kt-12 I've opened [this feature request](https://github.com/orgs/elementor/discussions/23621) on the Elementor repo. Once you're ready, can you open a new PR for this branch into their repo and reference that it currently is meant to address **Item 1 – Apply default optimization strategies during content rendering**?

kt-12 committed:
remove leftover action

arielk committed:
Merge branch 'main' into feature/img-lo-wp-logic-to-elementor

joemcgill committed:
Adjust formatting and CS

joemcgill committed:
Merge branch 'main' into feature/img-lo-wp-logic-to-elementor

joemcgill committed:
PHPUnit: Add group annotations

This allows unit tests for the specific image loading optimization module to be run by passing `--group image-loading-optimization` to the CLI.

joemcgill committed:
PHPUnit: Add setup for shared fixtures for image markup

joemcgill committed:
Merge branch 'main' into feature/img-lo-wp-logic-to-elementor

louiswol94 committed:
Merge branch 'main' into feature/img-lo-wp-logic-to-elementor

joemcgill committed:
Use short array syntax

Co-authored-by: Louis Wolmarans <louiswol94@gmail.com>

joemcgill committed:
Improve formatting of translated description

Co-authored-by: Rami Yushuvaev <92088692+rami-elementor@users.noreply.github.com>

joemcgill committed:
Add line break for readability

Co-authored-by: Louis Wolmarans <louiswol94@gmail.com>

joemcgill committed:
Abstract conditional logic to a variable for readability

joemcgill committed:
Update module description and release status.

joemcgill committed:
Allow minimum pixel threshold to be filterable

joemcgill committed:
Fix is_admin check in unit tests

joemcgill committed:
Add inline docs about copied logic

louiswol94 committed:
Merge branch 'main' into feature/img-lo-wp-logic-to-elementor

joemcgill committed:
CS Fixes from code review

Co-authored-by: Louis Wolmarans <louiswol94@gmail.com>

joemcgill committed:
Improve unit teststo be less brittle
